mirror.py

```diff
@@ -0,0 +1,70 @@
+import re
+import subprocess
+import typing
+from pathlib import Path
+
+import tomli
+import urllib3
+from packaging.requirements import Requirement
+from packaging.version import Version
+
+
+def main():
+    with open(Path(__file__).parent / "pyproject.toml", "rb") as f:
+        pyproject = tomli.load(f)
+
+    all_versions = get_all_versions()
+    current_version = get_current_version(pyproject=pyproject)
+    target_versions = [v for v in all_versions if v > current_version]
+
+    for version in target_versions:
+        paths = process_version(version)
+        subprocess.run(["git", "add", *paths])
+        subprocess.run(["git", "commit", "-m", f"Mirror: {version}"])
+        subprocess.run(["git", "tag", f"v{version}"])
+
+
+def get_all_versions() -> typing.List[Version]:
+    response = urllib3.request("GET", "https://pypi.org/pypi/ruff/json")
+    if response.status != 200:
+        raise RuntimeError("Failed to fetch versions from pypi")
+
+    versions = [Version(release) for release in response.json()["releases"]]
+    return sorted(versions)
+
+
+def get_current_version(pyproject: dict) -> Version:
+    requirements = [Requirement(d) for d in pyproject["project"]["dependencies"]]
+    requirement = next((r for r in requirements if r.name == "ruff"), None)
+    assert requirement is not None
```

Comment:
Should we include an assertion message that would help us diagnose this failed assertion?

Reply:
I added some assertion messages, but not sure if I need to supplement.